### PR TITLE
Support different schemas

### PR DESCRIPTION
**Pull request by SurajBDeore**
Add support to `lobster-codebeamer` to generate output using the following schemas:
- requirement
- implementation
- activity

The user can select the schema with a command line flag,
or through the configuration file.

Issue: https://github.com/bmw-software-engineering/lobster/issues/86

### CODE REVIEW

**Comment by mugdhadhole1** [comment]
I could see, the CI has failed while executing linter.
Always make sure you run make lint locally before you make a commit and push the code

File: lobster/tools/codebeamer/codebeamer.py [hunk 1]
```diff
@@ -45,17 +45,18 @@
...
-from lobster.items import Tracing_Tag, Requirement
+from lobster.items import Tracing_Tag, Requirement, Implementation, Activity
 from lobster.location import Codebeamer_Reference
 from lobster.errors import Message_Handler, LOBSTER_Error
 from lobster.io import lobster_read, lobster_write
 
 
 class References(Enum):
     REFS = "refs"
+    SCHEMA = "schema"
```
Comment (phiwuu) [inline lobster/tools/codebeamer/codebeamer.py:56]: Why did you add `SCHEMA` here?

**Review by phiwuu — CHANGES_REQUESTED** [review]

File: lobster/tools/codebeamer/codebeamer.py (hunk 1 above)
Comment (SurajBDeore) [reply lobster/tools/codebeamer/codebeamer.py:56]: > Why did you add `SCHEMA` here?

If we don't provide the supported reference, it will throw a KeyError.

File: lobster/tools/codebeamer/codebeamer.py [hunk 2]
```diff
@@ -237,9 +237,55 @@ def to_lobster(cb_config, cb_item):
...
+            tracker=cb_item["tracker"]["id"],
+            item=cb_item["id"],
+            version=cb_item["version"],
+            name=item_name
+        ),
+        'kind': kind
+    }
+
+
+def _create_lobster_item(schema_class, common_params, item_name, status):
```
Comment (phiwuu) [inline lobster/tools/codebeamer/codebeamer.py:264]: please add a unit test

File: lobster/tools/codebeamer/codebeamer.py [hunk 3]
```diff
@@ -237,9 +237,55 @@ def to_lobster(cb_config, cb_item):
                     continue
 
                 (map_reference_name_to_function[reference_name]
-                 (req, flat_values_list))
+                 (item, flat_values_list))
+
+    return item
+
+
+def _create_common_params(namespace: str, cb_item: dict, cb_root: str, item_name: str, kind: str):
```
Comment (phiwuu) [inline lobster/tools/codebeamer/codebeamer.py:245]: please add a unit test

**Review by phiwuu — CHANGES_REQUESTED** [review]
please test our changes :-)

File: lobster/tools/codebeamer/codebeamer.py [hunk 4]
```diff
@@ -172,7 +174,20 @@ def get_query(mh, cb_config, query_id):
 
     return rv
 
+def get_schema_config(cb_config):
```
Comment (kedarnn) [inline lobster/tools/codebeamer/codebeamer.py:177]: Please add a Docstring for the function. Something like below.
```python
"""The functions returns a schema map based on the schema mentioned in the cb_config dictionary. If there is no match then it results in an KeyError.

Positional arguments:
cb_config --  configuration dictionary containing the schema.
"""
```

File: lobster/tools/codebeamer/codebeamer.py [hunk 5]
```diff
@@ -172,7 +174,20 @@ def get_query(mh, cb_config, query_id):
 
     return rv
```
Comment (kedarnn) [inline lobster/tools/codebeamer/codebeamer.py:176]: Add one more blank line here. Classed and top-level functions should be separated by 2 blank lines.

File: lobster/tools/codebeamer/codebeamer.py [hunk 6]
```diff
@@ -237,9 +243,53 @@ def to_lobster(cb_config, cb_item):
                     continue
 
                 (map_reference_name_to_function[reference_name]
-                 (req, flat_values_list))
+                 (item, flat_values_list))
+
+    return item
+
+
+def _create_common_params(namespace: str, cb_item: dict, cb_root: str, item_name: str, kind: str):
```
Comment (kedarnn) [inline lobster/tools/codebeamer/codebeamer.py:251]: Same as [comment](https://github.com/bmw-software-engineering/lobster/pull/79/files#r1802539397)

File: lobster/tools/codebeamer/codebeamer.py [hunk 7]
```diff
@@ -237,9 +243,53 @@ def to_lobster(cb_config, cb_item):
...
+            item=cb_item["id"],
+            version=cb_item["version"],
+            name=item_name
+        ),
+        'kind': kind
+    }
+
 
-    return req
+def _create_lobster_item(schema_class, common_params, item_name, status):
```
Comment (kedarnn) [inline lobster/tools/codebeamer/codebeamer.py:270]: Same as [comment](https://github.com/bmw-software-engineering/lobster/pull/79/files#r1802539397)

File: test-unit/lobster-codebeamer/test_codebeamer_schema.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+            'common_params': common_params,
+            'item_name': item_name,
+            'expected_class': expected_class,
+            'tag': Tracing_Tag(namespace, str(cb_item["id"]), cb_item["version"]),
+            'location': Codebeamer_Reference(self.root_url, cb_item["tracker"]["id"], cb_item["id"], cb_item["version"], item_name),
+            'kind' : kind
+        }
+
+    def  generate_test_case(self):
+
```
Comment (kedarnn) [inline test-unit/lobster-codebeamer/test_codebeamer_schema.py:39]: No need to add a blank line at the start of the function. Please check for other functions and classes as well.

File: test-unit/lobster-codebeamer/test_codebeamer_schema.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+            self.generate_common_params('imp', 'Implementation Item', 'implementation', Implementation),
+            self.generate_common_params('act', 'Activity Item', 'activity', Activity),
+            ]
+
+    def test_create_common_params(self):
+
+        test_cases = self.generate_test_case()
+        for case in test_cases:
+            with self.subTest(case=case):
+
```
Comment (kedarnn) [inline test-unit/lobster-codebeamer/test_codebeamer_schema.py:51]: No need for a blank line.

File: lobster/tools/codebeamer/codebeamer.py [hunk 10]
```diff
@@ -319,6 +372,11 @@ def main():
                     default=30,
                     help="Timeout in s (by default 30) for each REST call.")
 
+    ap.add_argument("--schema",
+                    default='requirement',
+                    help="Specify the output schema (Requirement, Implementation, Activity)."
+                    )
```
Comment (kedarnn) [inline lobster/tools/codebeamer/codebeamer.py:378]:
```suggestion
    ap.add_argument("--schema",
                    default='requirement',
                    help="Specify the output schema (Requirement, Implementation, Activity).")
```

**Comment by mugdhadhole1** [comment]
I would really suggest to look at the linting errors.
I could see `make lint` showed a lot of linting errors

File: packages/lobster-tool-codebeamer/README.md [hunk 11]
```diff
@@ -11,31 +11,50 @@ requirements management tool
...
+  1. Activity: Sets the schema to lobster-act-trace.
+  2. Implementation: Sets the schema to lobster-imp-trace.
+  3. Requirement: Sets the schema to lobster-req-trace.
+
+  If the kind is not specified, the tool will default to Requirement, and the schema lobster-req-trace will be used.
+
+  Here’s an example configuration:
+  ```json
+  {
+    "kind": "Activity",  // Specifies schema
```
Comment (mugdhadhole1) [inline packages/lobster-tool-codebeamer/README.md:52]: Here, I suppose it should be "schema"
because, If I add "kind" and "refs" to the config file it gives an error message - 

> KeyError: "The provided config keys are not supported! supported keys: 'refs, schema'"

**Comment by mugdhadhole1** [comment]
Currently, the command accepts both "--schema" as well as "config file."
Hence, it creates ambiguity in fetching "schema"
Suggestion, we should either use --schema flag or schema in config file

File: packages/lobster-tool-codebeamer/README.md (hunk 11 above)
Comment (mugdhadhole1) [inline packages/lobster-tool-codebeamer/README.md:52]:
```suggestion
    "schema": "Activity",  // Specifies schema
```

Here, I suppose it should be "schema"
because, If I add "kind" and "refs" to the config file it gives an error message - 

> KeyError: "The provided config keys are not supported! supported keys: 'refs, schema'"

**Comment by SurajBDeore** [comment]
> please test our changes :-)

Tested

**Review by kedarnn — APPROVED** [review]
Approved

**Review by mugdhadhole1 — APPROVED** [review]

**Review by phiwuu — APPROVED** [review]

**Review by phiwuu — APPROVED** [review]
